Remove old rendering backend.

glennw opened the pull request:
This removes paint threads, rust-layers dependency, and changes
optional webrender types to be required.

The use_webrender option has been removed, however I've left
the "-w" command line option in place for now so that wpt
runner can continue to pass that. Once it's removed from there
we can also remove the -w option.

Once this stage is complete, it should be fine to change the
display list building code to generate webrender display
lists directly and avoid the conversion step.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13711)

<!-- Reviewable:end -->

gw3583 committed:
Remove old rendering backend.

This removes paint threads, rust-layers dependency, and changes
optional webrender types to be required.

The use_webrender option has been removed, however I've left
the "-w" command line option in place so that wpt
runner can continue to pass that. Once it's removed from there
we can also remove the -w option.

Once this stage is complete, it should be fine to change the
display list building code to generate webrender display
lists directly and avoid the conversion step.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/constellation/lib.rs, components/constellation/pipeline.rs, components/constellation/Cargo.toml, components/constellation/constellation.rs
 * @kichjang: components/script/dom/canvasrenderingcontext2d.rs, components/script/dom/window.rs, components/script_traits/lib.rs, components/script_traits/lib.rs, components/script_layout_interface/Cargo.toml, components/script_traits/script_msg.rs, components/script_traits/script_msg.rs, components/net/image_cache_thread.rs, components/script_layout_interface/lib.rs, components/script/script_thread.rs, components/script_layout_interface/message.rs, components/net_traits/image/base.rs, components/net_traits/image/base.rs, components/script_traits/Cargo.toml, components/script_traits/Cargo.toml

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify gfx, layout, net, and script code, but no tests are modified. Please consider adding a test!

glennw commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit ba9d0bb with merge e429ebd...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/232)

glennw commented:
@bors-servo retry

glennw commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit c241737 with merge 409399e...

bors-servo commented:
:broken_heart: Test failed - [windows-dev](http://build.servo.org/builders/windows-dev/builds/1441)

Ms2ger commented on components/servo/Cargo.lock:
```diff
@@ -2693,6 +2658,11 @@ dependencies = [
 name = "webrender"
 version = "0.6.0"
 source = "git+https://github.com/servo/webrender#b31b4cf76324bfbe82d27a059e1330ef5ae34c35"
+replace = "webrender 0.6.0"
```
Huh?

glennw commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 90c3764 with merge e43bd42...

glennw commented:
We should get the stacking context patch from @mrobinson reviewed and merged before this. This can still be reviewed, and I'll rebase on top of that when it lands.

emilio commented on components/canvas/canvas_paint_thread.rs:
```diff
@@ -542,14 +542,12 @@ impl<'a> CanvasPaintThread<'a> {
+                                            webrender_traits::ImageFormat::RGBA8,
+                                            element.into());
 
             let pixel_data = CanvasPixelData {
```
Please fill a follow-up to remove the `image_data` field from `CanvasPixelData`.

glennw replied:
https://github.com/servo/servo/issues/13734

emilio commented on components/compositing/compositor.rs:
```diff
@@ -756,18 +614,7 @@ impl<Window: WindowMethods> IOCompositor<Window> {
             }
 
             (Msg::CollectMemoryReports(reports_chan), ShutdownState::NotShuttingDown) => {
```
At this point we could also remove this message entirely. Please do it or file a followup for it.

glennw replied:
https://github.com/servo/servo/issues/13735

emilio commented on components/compositing/compositor.rs:
```diff
@@ -1213,65 +765,11 @@ impl<Window: WindowMethods> IOCompositor<Window> {
-        }
+                                _pipeline_id: PipelineId,
+                                _layer_id: LayerId,
+                                _point: Point2D<f32>) {
```
So if I understand this code correctly, this is only called on resize to keep the scroll where the fragment of the URL points to. The code in `script_thread.rs` has a `FIXME(pcwalton)` saying it's very bogus. I think it's a regression we could take, what do you think?

glennw replied:
It's also used for scrolling to a # anchor in a page. However, it doesn't work with webrender right now anyway, so it's not technically a regression. https://github.com/servo/servo/issues/13736

emilio reviewed (commented):
Only a few comments, not an in-depth review.

glennw commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 29567a8 with merge 1102593...

glennw commented on components/compositing/compositor.rs:
```diff
@@ -229,24 +200,14 @@ pub struct IOCompositor<Window: WindowMethods> {
-    webrender: Option<webrender::Renderer>,
+    webrender: webrender::Renderer,
 
     /// The webrender interface, if enabled.
```
Fixed

glennw commented on components/compositing/compositor.rs:
```diff
@@ -358,26 +306,9 @@ fn initialize_png(width: usize, height: usize) -> RenderTargetInfo {
 
     gl::bind_texture(gl::TEXTURE_2D, 0);
 
-    let renderbuffer_ids = if opts::get().use_webrender {
```
That piece of code was WR-only, but it's now not required by WR at all - so I've removed it completely.

glennw commented on components/compositing/compositor_thread.rs:
```diff
@@ -8,16 +8,14 @@ use SendableFrameTree;
 use euclid::point::Point2D;
 use euclid::size::Size2D;
-use gfx_traits::{Epoch, FrameTreeId, LayerId, LayerProperties, PaintListener};
+use gfx_traits::{LayerId};
```
Fixed

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3573)

glennw commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 6ffab26 with merge c8fbd74...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3499), [arm64](http://build.servo.org/builders/arm64/builds/3511), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3580), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/206), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/206), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3576), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3555), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/257), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/258), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1464)

glennw commented:
This passes bors try run now, and I've addressed the comments so far. r? @larsbergstrom @metajack @pcwalton and anyone else who has time to look over some of these changes.

larsbergstrom commented:
Reviewed 51 of 53 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[components/compositing/compositor.rs, line 771 at r1](https://reviewable.io:443/reviews/servo/servo/13711#-KTrTZZr0Z4fhEPNXpKG-r1-771:-KTvcIsmSbEdQxD_msgs:b-86xeb2) ([raw file](https://github.com/servo/servo/blob/c233541c13bf792afc2691d4df745bd1f5507a33/components/compositing/compositor.rs#L771)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary>
> It's also used for scrolling to a # anchor in a page. However, it doesn't work with webrender right now anyway, so it's not technically a regression. https://github.com/servo/servo/issues/13736
</details>
Hrm, 

---

*[ports/cef/window.rs, line 298 at r2](https://reviewable.io:443/reviews/servo/servo/13711#-KTveqU9XxzQo4m4CsPD:-KTveqU9XxzQo4m4CsPE:b-rhejiz) ([raw file](https://github.com/servo/servo/blob/768a699fa5d822173b2247d6348413d09951a8bb/ports/cef/window.rs#L298)):*
> ```Rust
>     fn native_display(&self) -> NativeDisplay {
>         NativeDisplay::new()
>     }
> ```

@discomfitor Do we need these methods for the Linux CEF bindings?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13711)*
<!-- Sent from Reviewable.io -->

metajack commented:
Let's open bugs for the `-w` flag removal and the fixing of scrolling to anchors.

pcwalton commented:
LGTM

glennw commented:
@metajack https://github.com/servo/servo/issues/13736 https://github.com/servo/servo/issues/13761

larsbergstrom commented:
@glennw Do you know if that last of an exposed binding will break CEF? Other than that and a rebase, it looks like this is g2g.

glennw commented:
@larsbergstrom Rebased. I don't believe that native_display() removal affects the cef build, if that's what you're asking?

emilio commented:
Yeah, that `native_display` method was only used to pass the appropriate X display to `rust-layers` AFAIK. In fact I think that now all the logic for that both in the CEF and glutin directories can be removed.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13756) made this pull request unmergeable. Please resolve the merge conflicts.

glennw commented:
@larsbergstrom Rebased again, over the clip changes :)

glennw commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 324914d with merge c57eee2...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/268)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/inline_block_opacity_change.html
      └   → /_mozilla/css/inline_block_opacity_change.html aed6845267bba6c52d3aa69c4889449b454d8117
    /_mozilla/css/inline_block_opacity_change_ref.html febec1b4730afab195f280694dad47200b09ea3c
    Testing aed6845267bba6c52d3aa69c4889449b454d8117 == febec1b4730afab195f280694dad47200b09ea3c

glennw commented:
@bors-servo retry

* #13360

bors-servo commented:
:hourglass: Trying commit 324914d with merge 577b37b...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3560), [arm64](http://build.servo.org/builders/arm64/builds/3571), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3637), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/269), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/269), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3638), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3616), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/318), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/320), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1527)

glennw commented:
@larsbergstrom Rebased and try passed, so should be good to go when you want to flick the switch :)

larsbergstrom commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 324914d has been approved by `larsbergstrom`

<!-- @bors-servo r=larsbergstrom 324914dd7339ecea2fc76f7f68447422f6fae6fc -->

larsbergstrom commented:
@bors-servo clean try- r+

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #13801

bors-servo commented:
:pushpin: Commit 324914d has been approved by `larsbergstrom`

<!-- @bors-servo r=larsbergstrom 324914dd7339ecea2fc76f7f68447422f6fae6fc -->

emilio commented:
@bors-servo: p=1

 * Prone to bitrot

emilio commented:
@bors-servo: clean r- try-

emilio commented:
@bors-servo: r=larsbergstrom,pcwalton,Ms2ger,emilio

bors-servo commented:
:pushpin: Commit 324914d has been approved by `larsbergstrom,pcwalton,Ms2ger,emilio`

<!-- @bors-servo r=larsbergstrom,pcwalton,Ms2ger,emilio 324914dd7339ecea2fc76f7f68447422f6fae6fc -->

glennw commented:
@bors-servo r=larsbergstrom

bors-servo commented:
:pushpin: Commit acfdfd2 has been approved by `larsbergstrom`

<!-- @bors-servo r=larsbergstrom acfdfd2fa98562c9f891fbee2dbc3424803c1a3c -->

bors-servo commented:
:hourglass: Testing commit acfdfd2 with merge f96718d...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3572), [arm64](http://build.servo.org/builders/arm64/builds/3583), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3649), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/282), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/283), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3650), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3628), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/330), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/332), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1539)
